### PR TITLE
Timebounded

### PR DESCRIPTION
**Pull request by nmattia**
Allow local specification of sampling configuration and add helper for running time bounded benchmarks.

### CODE REVIEW

File: src/Hyperion/Benchmark.hs [hunk 1]
```diff
@@ -29,12 +30,14 @@ import Data.Text (Text)
...
 import Control.DeepSeq
 import Hyperion.Internal
+import Hyperion.Measurement
 
 data Benchmark where
   Bench :: Text -> Batch () -> Benchmark
   Group :: Text -> [Benchmark] -> Benchmark
   Bracket :: NFData r => IO r -> (r -> IO ()) -> (Env r -> Benchmark) -> Benchmark
   Series :: Show a => Vector a -> (Env a -> Benchmark) -> Benchmark
+  WithSampling :: (Batch () -> IO Sample) -> Benchmark -> Benchmark
```
Comment (mboes) [inline src/Hyperion/Benchmark.hs]: Why do we need an extra constructor to the benchmark DSL?

File: src/Hyperion/Benchmark.hs [hunk 2]
```diff
@@ -58,6 +63,9 @@ bgroup name bks = Group (Text.pack name) bks
 series :: Show a => Vector a -> (Env a -> Benchmark) -> Benchmark
 series = Series
 
+withSampling :: (Batch () -> IO Sample) -> Benchmark -> Benchmark
```
Comment (mboes) [inline src/Hyperion/Benchmark.hs]: I don't know what this function does... it has no haddock.

**Comment by nmattia** [comment]
This allows the user to specify the sampling strategy. It also adds a function helper to specify a sampling strategy with a bound on the total benchmark time (for a benchmark branch).

The reason for wanting to configure the sampling strategy is that some benchmarks take too much time if the operation being measured goes through the 500 iterations by default. For instance the `hyperion-example` executable on master takes over a minute to run on my machine. Also it is not clear (to me at least) whether we always want _more samples_ or _more time_ when it comes to homogenizing the measurements.  

The current master already allows to specify a sampling strategy, but:
1. it's hidden inside `runBenchmarkWithConfig` and is not accessible through `defaultMain`, `doRun`, etc.
1. it only allows to specify a strategy for the whole benchmark tree, not for specific `bench`s.

The reason for wanting to tune the benchmark strategy locally is that all benchmarks do not have the same requirements. Some benchmarks might need a few thousand iterations to be precise, some don't. Allowing a given branch to have a specific strategy means that the trade-offs are local, and the same tree can hold various benchmarks (all rooted in the same `defaultMain`). 

By passing the sampling strategy in the recursive `go` function of `runBenchmarkWithConfig` we allow the reader `(->) (Batch () -> IO Sample)` to use a `local` equivalent when traversing the tree. Adding the `WithSampling` constructor allows hyperion to observe the different strategies while walking the `Benchmark` structure, which can be influenced by the API user.

**Comment by nmattia** [comment]
Addresses #18

**Pull request merged**